### ISSUE TITLE
OnPush change detection for dynamically created components

### ISSUE THREAD

**Issue by ChrisFulstow** [angular/angular#14087]
**I'm submitting a ...**
```
[x] bug report
[ ] feature request
[ ] support request
```

**Current behavior**
Components added dynamically that have `ChangeDetectionStrategy.OnPush` do not detect input changes. For example, if such a component is instantiated at runtime via `ViewContainerRef.createComponent`, then setting an `@Input` property to a new immutable object does not trigger change detection. Instead, the component must be marked for check manually, with `ChangeDetectorRef.markForCheck`, rather than triggering change detection automatically.

**Expected behavior**
Setting an `@Input` property value on a dynamically created component that has `OnPush` change detection should automatically trigger a change detection run, following the same behavior as components created from a template.

**Minimal reproduction of the problem with instructions**
[Plunker to demonstrate](https://plnkr.co/edit/0wHQghtww2HXVbC27bC1)
The ChangeDetectionOnPushComponent does not detect the change to its @Input property.

(Also [related Stack Overflow question](http://stackoverflow.com/questions/41797590/angular-2-onpush-change-detection-for-dynamic-components) with markForCheck workaround.)

**What is the motivation / use case for changing the behavior?**
I believe this may be a bug because push change detection for dynamically created components behaves differently from static components, and is not what might be expected for `ChangeDetectionStrategy.OnPush`. My use case is an application that uses many dynamically created nested components that would benefit from lightweight on-push change detection.

* **Angular version:** 2.0.X
2.4.4

* **Browser:**
All
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.0.2

**Comment by DzmitryShylovich** [comment]
To me it works as expected. 
> this.component.instance.message = this.message;

you are just setting a class property. how angular can know about that change and trigger cd?

**Comment by ChrisFulstow** [comment]
True but the class is a component and the property is an `@Input` property, so for consistency with template components the change should trigger a change detection run. Here, Angular detects the component input change when the change detection strategy is `Default`, so it should also detect eligible input changes when the strategy is `OnPush`.

**Comment by DzmitryShylovich** [comment]
> @Input property

it's just an annotation. it doesn't do anything.

> template components the change should trigger a change detection

for "template" components angular generated host component (a wrapper) which handles things like cd etc.

And in your case it's just a regular js class without any wrappers. There's no magic. 
The only way I see is to create a setter from a property somehow...

**Comment by ChrisFulstow** [comment]
Agreed, creating a setter on the input property to include a `markForCheck()` is a good solution.

But I still feel maybe there is a bug here because change detection _does_ run on the component when it has ChangeDetectionStrategy.Default, but not when it has ChangeDetectionStrategy.OnPush.

When the component is created by ViewContainerRef.createComponent, doesn't that also create the ComponentRef wrapper containing a ChangeDetectorRef for the new component?

**Comment by shlomiassaf** [comment]
@DzmitryShylovich @ChrisFulstow this is an expected behaviour, it's just very hard to explain and understand due to all of the internals that run behind as well as the complex component parent-child structure. I will do my best...

To make it more clear I changed the demo a bit  https://plnkr.co/edit/lZlyl072R1XXgifwTUaq?p=preview

Took both components (defulat, onpush) and duplicated them again, this time in the template so angular will instantiate them.

Results, all components react to change except dynamic OnPush component

The components defined on the template are fired since their parent change detector is doing logic to check for changes on them as well. this is done by calling all `@Input` checking functions, `check_message` in our case, on the component as well as running `ngDoCheck` on it.

`ngDoCheck` will check if there was a change, if so it will return true (check_message will make sure it will return true), then the container component will make sure to change the change detection mode of the component to `CheckOne` , then the call to check for changes (internal) on the component will fire.

The `default` components have the change detection mode set to `CheckAlways` so they always hit.

The dynamic component is not connected to the container, so the container does not handle that logic. The change detection logic runs but it does not go through to actually changing anything b because the change detection mode is `Checked`, remind you that it was checked for the templated components but was changed by it's container.

So when you create a dynamic component, you must hook it into the change detection process when you `OnPush`.

When using templates you get it for free by angular, when you go dynamic you must do it manually.
